File: maximum-length-of-a-concatenated-string-with-unique-characters.rs
use std::collections::HashSet;

impl Solution {
  pub fn max_length(arr: Vec<String>) -> i32 {
    let n = arr.len();
    let mut ans = 0;
    for state in 0..(1 << n) {
      let mut s = String::new();
      for i in 0..n {
        if state & (1 << i) != 0 {
          s.push_str(&arr[i]);
        }
      }
      if s.len() != s.chars().collect::<HashSet<char>>().len() {
        continue;
      }

      ans = ans.max(s.len() as i32);
    }
    ans
  }
}

File: maximum-length-of-a-concatenated-string-with-unique-characters.cc
class Solution {
public:
  int maxLength(vector<string>& arr) {
    int N = arr.size(), ans = 0, B[26]{};
    for (int s = 0; s < (1 << N); ++s) {
      for (int i = 0; i < N; ++i) {
        if (!(s & (1 << i))) continue;
        for (char& c : arr[i]) ++B[c - 'a'];
      }

      int cur = 0, ok = 1;
      for (int l = 0; l < 26; ++l) {
        if (B[l] > 1) ok = 0;
        else cur += B[l];
      }
      if (ok) ans = max(ans, cur);

      for (int i = 0; i < N; ++i) {
        if (!(s & (1 << i))) continue;
        for (char& c : arr[i]) --B[c - 'a'];
      }
    }

    return ans;
  }
};
